### PR TITLE
feat: AWS Secrets Manager Vault implementation

### PR DESCRIPTION
**Pull request by FlorianSeidelAWS**
## What this PR changes/adds

The PR adds a new extension module which implements the Vault interface using AWS SecretsManager. It does not depend on other AWS related modules and uses the default credentials chain to resolve credentials. This means it can be used as a drop-in replacement for AWS deployments. 

## Why it does that

AWS deployments of EDC can now use Secrets Manager instead of a self managed vault, which reduces cost and management overhead.

## Further notes

- Secret keys are sanitized to comply with AWS Secrets Manager key requirements.

## Linked Issue(s)

-

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

File: extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultException.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/*
+ *  Copyright (c) 2022 Amazon Web Services
```
Comment (jimmarino) [inline extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultException.java:2]: Nit: should this be 2022-2023 or just 2023?

File: extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultExtension.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.eclipse.edc.spi.security.VaultPrivateKeyResolver;
+import org.eclipse.edc.spi.system.ServiceExtension;
+import org.eclipse.edc.spi.system.ServiceExtensionContext;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
+
+import static org.eclipse.edc.util.configuration.ConfigurationFunctions.propOrEnv;
+import static org.eclipse.edc.util.string.StringUtils.isNullOrEmpty;
+
+@Provides({ Vault.class, PrivateKeyResolver.class, CertificateResolver.class })
```
Comment (jimmarino) [inline extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultExtension.java:33]: Could you add some documentation here?

File: extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVault.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @param originalKey any key
+     * @return Valid AWS Secrets Manager key
+     */
+    String sanitizeKey(String originalKey) {
+        var key = originalKey;
+        if (originalKey.length() > 500) {
+            key = originalKey.substring(0, 500);
+        }
+        StringBuilder sb = new StringBuilder();
```
Comment (jimmarino) [inline extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVault.java:118]: super nit, for consistency I would use "var" here and below

File: extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVault.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        for (int i = 0; i < key.length(); i++) {
+            Character c = key.charAt(i);
+            if (!Character.isLetterOrDigit(c) && c != '/' && c != '_' && c != '+' && c != '.' && c != '@' && c != '-') {
+                replacedIllegalCharacters = true;
+                sb.append('-');
+            } else {
+                sb.append(c);
+            }
+        }
+        String newKey = sb.append('_').append(originalKey.hashCode()).toString();
```
Comment (jimmarino) [inline extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVault.java:129]: Nit: please use var

**Review by jimmarino — CHANGES_REQUESTED** [review]
Nice! Just a few small change requests

File: extensions/common/vault/vault-aws/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+dependencies {
+    api(project(":spi:common:core-spi"))
+    implementation("software.amazon.awssdk:secretsmanager:2.19.3")
+    implementation(project(":core:common:util"))
+    testImplementation(libs.mockito.inline)
+}
+
+
+publishing {
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/build.gradle.kts:27]: the whole block can be removed as the publication is already added by the `edc-build` plugin

File: extensions/common/vault/vault-aws/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+plugins {
+    `java-library`
+}
+
+dependencies {
+    api(project(":spi:common:core-spi"))
+    implementation("software.amazon.awssdk:secretsmanager:2.19.3")
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/build.gradle.kts:21]: please provide a PR on the `GradlePlugins` repository to add this library version to the version-catalog so we can use it here.
https://github.com/eclipse-edc/GradlePlugins/blob/main/gradle/libs.versions.toml

File: extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            var validKey = "valid" + validCharacter + "key";
+            assertThat(vault.sanitizeKey(validKey)).isEqualTo(validKey + '_' + validKey.hashCode());
+        }
+        var key2 = "invalid#key";
+        var sanitized = vault.sanitizeKey(key2);
+        assertThat(sanitized).isEqualTo("invalid-key" + "_" + key2.hashCode());
+    }
+
+    @Test
+    void resolveSecret_sanitizeKeyNameLimitsKeySize() {
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java:51]: please use empty lines to separate "arrange" "act" and "assert" phases of the test

File: extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultExtension.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                .region(Region.of(vaultRegion));
+        return builder.build();
+    }
+
+    private String getMandatorySetting(ServiceExtensionContext context, String setting) {
+        var value = context.getSetting(setting, null);
+        if (isNullOrEmpty(value)) {
+            value = propOrEnv(setting, null);
+            if (isNullOrEmpty(value)) {
+                throw new AwsSecretsManagerVaultException(String.format("'%s' must be supplied but was null", setting));
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/src/main/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultExtension.java:75]: please use `EdcException` instead, no need to define a custom one.

File: extensions/common/vault/vault-aws/src/main/resources/META-INF/services/org.eclipse.edc.spi.system.ServiceExtension [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#
+#  Copyright (c) 2020, 2021, 2022 Amazon Web Services Corporation
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/src/main/resources/META-INF/services/org.eclipse.edc.spi.system.ServiceExtension:2]: I'm not a copyright expert, but I think just a single 2023 could be enough

**Review by ndr-brt — CHANGES_REQUESTED** [review]

File: extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        assertThat(sanitized.length()).isLessThanOrEqualTo(512);
+    }
+
+
+    @Test
+    void storeSecret_shouldSanitizeKey() {
+        var key = "invalid#key";
+        var value = "value";
+        vault.storeSecret(key, value);
+        Mockito.verify(secretClient).createSecret(CreateSecretRequest.builder().name(vault.sanitizeKey(key))
```
[the remark this answers is not in the document]
Comment (FlorianSeidelAWS) [reply extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java]: This actually checks if the key used to store the secret is the sanitized key. I'll leave it like this.

**Comment by paullatzelsperger** [comment]
removed myself again, since @jimmarino and @ndr-brt seem to have this covered :)

File: settings.gradle.kts [hunk 11]
```diff
@@ -41,6 +41,9 @@ dependencyResolutionManagement {
             from("org.eclipse.edc:edc-versions:0.0.1-SNAPSHOT")
             // this is not part of the published EDC Version Catalog, so we'll just "amend" it
             library("dnsOverHttps", "com.squareup.okhttp3", "okhttp-dnsoverhttps").versionRef("okhttp")
+            // Version ref from version catalog is not recognized for aws. Reasons are not clear.
+            // TODO: Remove this once secrets manager is part of the EDC version catalog.
+            library("aws.secretsmanager", "software.amazon.awssdk", "secretsmanager").version("2.19.26")
```
Comment (ndr-brt) [inline settings.gradle.kts:46]: I tried it and the entry in the version catalog now works, so this isn't needed

File: extensions/common/vault/vault-aws/build.gradle.kts [hunk 12]
```diff
@@ -18,17 +18,7 @@ plugins {
 
 dependencies {
     api(project(":spi:common:core-spi"))
-    implementation("software.amazon.awssdk:secretsmanager:2.19.3")
+    api(libs.aws.secretsmanager)
```
Comment (ndr-brt) [inline extensions/common/vault/vault-aws/build.gradle.kts:21]:
```suggestion
    implementation(libs.aws.secretsmanager)
```

**Review by ndr-brt — CHANGES_REQUESTED** [review]

File: extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java [hunk 13]
```diff
@@ -17,111 +17,94 @@
...
 
     @Test
     void storeSecret_shouldSanitizeKey() {
         var key = "invalid#key";
         var value = "value";
+
         vault.storeSecret(key, value);
-        Mockito.verify(secretClient).createSecret(CreateSecretRequest.builder().name(vault.sanitizeKey(key))
+
+        verify(secretClient).createSecret(CreateSecretRequest.builder().name(sanitizer.sanitizeKey(key))
```
[the remark this answers is not in the document]
Comment (FlorianSeidelAWS) [reply extensions/common/vault/vault-aws/src/test/java/org/eclipse/edc/vault/aws/AwsSecretsManagerVaultTest.java]: I replaced the sanitizer with a mock. Hope this makes more sense.

File: settings.gradle.kts (hunk 11 above)
Comment (FlorianSeidelAWS) [reply settings.gradle.kts]: Removed the lines. Works indeed. Thanks!

File: extensions/common/vault/vault-aws/build.gradle.kts (hunk 12 above)
Comment (FlorianSeidelAWS) [reply extensions/common/vault/vault-aws/build.gradle.kts]: Thanks. Didn't know the difference between api and implementation. The more you know...

**Review by ndr-brt — APPROVED** [review]
Good to go!

**Review by jimmarino — APPROVED** [review]

**Pull request merged**